crates/subspace-service/src/dsn/import_blocks/root_blocks.rs

```diff
@@ -0,0 +1,177 @@
+                            },
+                        )
+                        .await;
+
+                    match request_result {
+                        Ok(RootBlockResponse { root_blocks }) => {
+                            trace!(%peer_id, "Last root block request succeeded.");
+
+                            let last_root_block = root_blocks
+                                .iter()
+                                .max_by(|rb1, rb2| compare_optional_root_blocks(rb1, rb2));
+
+                            if let Some(Some(root_block)) = last_root_block {
+                                trace!(
+                                    %peer_id,
+                                    segment_index=root_block.segment_index(),
+                                    "Last root block was obtained."
+                                );
+
+                                return Ok(*root_block);
+                            } else {
+                                debug!(%peer_id, "Last root block was not received.");
+                            }
+                        }
+                        Err(error) => {
+                            debug!(%peer_id, ?error, "Last root block request failed.");
+                        }
+                    };
+                }
+                Err("No more peers for root blocks.".into())
+            }
+            Err(err) => {
+                warn!(?err, "get_closest_peers returned an error");
+
+                Err(err.into())
+            }
+        }
+    }
+
+    async fn get_root_blocks_batch(
+        &self,
+        segment_indexes: Vec<SegmentIndex>,
+    ) -> Result<Vec<Option<RootBlock>>, Box<dyn Error>> {
+        trace!(?segment_indexes, "Getting root block batch...");
+
+        // Get random peers. Some of them could be bootstrap nodes with no support for
+        // request-response protocol for records root.
+        let get_peers_result = self
+            .dsn_node
+            .get_closest_peers(PeerId::random().into())
+            .await;
+
+        match get_peers_result {
+            Ok(mut get_peers_stream) => {
+                while let Some(peer_id) = get_peers_stream.next().await {
+                    trace!(%peer_id, "get_closest_peers returned an item");
+
+                    let request_result = self
+                        .dsn_node
+                        .send_generic_request(
+                            peer_id,
+                            RootBlockRequest::SegmentIndexes {
+                                segment_indexes: segment_indexes.clone(),
+                            },
+                        )
+                        .await;
+
+                    match request_result {
+                        Ok(RootBlockResponse { root_blocks }) => {
+                            trace!(%peer_id, ?segment_indexes, "Root block request succeeded.");
+
+                            return Ok(root_blocks);
```

Comment:
Yes, I added TODOs with peer bans for the later improvements. Please, let me know if you want to have it in this PR.

Reply:
Since we should have basic checks that root blocks form a chain and so on, it should be sufficient to return peer ID from here and call `node.ban_peer()`, so I'd say yet, we want to have that from the very beginning if it is indeed that simple.

Also as a general rule about architecture if you know you'll be banning peers eventually, the API should return peer ID from here if it is not used initially (with corresponding TODO in the code).